### PR TITLE
First take at Dagger CLI

### PR DESCRIPTION
**Pull request by marcosnils**
Only adds two sub-commands to run GraphQL queries and start the engine
API router initially.

Current task list (offload) after deciding where the CLI will live: 
- [ ] merge `publish-cli.yml` & `publish-engine.yml` into `publish-engine-and-cli.yml`
- [ ] only publish the CLI when we are publishing a tag, e.g. `v0.3.5`
- [ ] `brew install dagger@0.3` - for [our tap formula](https://github.com/dagger/homebrew-tap)
  - Open question: **does Homebrew core allow this for bottled formulas?**
- [ ] validate env vars in `cli:publish`, e.g. [`sdk:nodejs:publish`](https://github.com/dagger/dagger/pull/3809/files#diff-403f05c48ac8fbf799afe4dcf233f44d126ed2546723938ac76b1e30cbb1c75cR54-R65)

Follow-up:
- version current Homebrew bottled formula: `dagger.rb` -> `dagger@0.2.rb`
- version current Homebrew tap formula: `dagger.rb` -> `dagger@0.2.rb`
  - it's OK to inconvenience the few users that use this - as discussed with @aluzzardi
- publish

Signed-off-by: Marcos Lilljedahl <marcosnils@gmail.com>

### CODE REVIEW

**Comment by gerhard** [comment]
Had a sync conversation with @marcosnils, this is what we are thinking cc @aluzzardi @sipsma @shykes @mircubed 

```mermaid
flowchart TB
    classDef fail fill:#f00;
    classDef success fill:#090;
    tag[Git tag >= v0.3.4] --> code[Check out code]

    code --> cli[mage cli:publish]
    subgraph GoReleaser
        cli --> goreleaser
        goreleaser --> |cli| github[GitHub release v0.3.x]
        goreleaser --> |cli| s3[AWS S3]
        goreleaser --> |cli| tap[Homebrew tap]
        goreleaser --> |engine reference| github
    end

    subgraph Dagger
        code --> engine[mage engine:publish]
        engine --> |engine| oci[ghcr.io/dagger/engine:v0.3.4]
    end
```

[Live graph editor](https://mermaid.live/edit#pako:eNqFUk1vnDAQ_SuWe2UJDrALRI3UdqXk0ku3UqXAHowZwArYyDZN0t397zWYTZpUSrgwH2_ePB5zwExWgDNcd_KBtVQZ9PNrIZB9WEe13kKNaso7VPOuyz7VQXD1pqtHxkDrBRCkZ4ChTX7DzfRG15_R78AP_WiPVqtrNK3Mv7XA7pEczZzuC7Hw2sSBOp73tIEpyIax7Lhu9w6jx7JRdGjRjfwBHVANyjWcMD7PN1L913upzZCjxR5Rw007lpPW27FEC8Dpfdx_MKvD_MuvHdqFHwENHfJb2UOp4GFK3sGDaLgAK6QGBYLBWaGbAFGdrXq2YUub5pUFZw8dlbPRxW-cdIzzvn92H5FkPG9apnwuL6qZ_WKZX_7jsxjs4R5UT3llj-gwlQtsWuihwJkNK6ruC1yIk8XR0cjdk2A4M2oED49DRQ1sObVf0eOspp221YGKOylf5Tg74EeckXXqJwmJ0jBOExKnwcbDTzhbBX4Qk8sgCskmXUfkMiHr-OThPzML8SMShantJBHZrOMo8jBU3Ej13R3-fP-nv5Uu8io)

**Comment by gerhard** [comment]
@marcosnils is this related to?
- https://github.com/dagger/dagger/issues/3653

**Comment by marcosnils** [comment]
> @marcosnils is this related to?

related but from the distance. The CLI will eventually use the CLI helper, but for the moment, according to what we spoke with @sipsma and @aluzzardi we'll keep using the legacy buildkit provisioner since it makes more sense at this stage.

**Comment by gerhard** [comment]
https://github.com/gerhard/dagger/releases/tag/v0.6.1 produced via https://github.com/gerhard/dagger/actions/runs/3447576047

https://github.com/gerhard/dagger/releases/tag/v0.5.1 produced locally via `./hack/make cli:publish`

**Comment by gerhard** [comment]
I noticed this morning that my fork created a `0.5.1` engine package. This should not have happened. I deleted this package version manually.

image

image

The PAT that I am using has too many privileges. On top of that, the workflow shouldn't have run at all. Mentioning it here since I do not have the bandwidth to tackle it straight away. Leaving it as a reminder for a quieter time.

File: cmd/dagger/listen.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+// func startEngine(conf *engine.Config) error {
+// return nil
+// }
```
Comment (gerhard) [inline cmd/dagger/listen.go:54]: Remove this and keep for another PR.

File: cmd/dagger/query.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+	result, err := json.MarshalIndent(res, "", "    ")
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func getKVInput(kvs []string) map[string]string {
```
Comment (sipsma) [inline cmd/dagger/query.go:119]: Just FYI this ability to provide vars as a command line flag has a bunch of problems, biggest one being that you can't really provide a list: https://github.com/dagger/dagger/issues/3082

We were okay with it at the time since everything was in prototype phase, but I think we should either fix this or remove it for now before the release.

**Comment by gerhard** [comment]
Current task list (offload):
- [ ] merge publish-cli.yml & publish-engine.yml into `publish-engine-and-cli.yml`
- [ ] only publish the CLI when we are publishing a tag
- [ ] `brew install dagger@0.3` - for the tap. Open question: does Homebrew core allow this?
- [ ] validate env vars in `cli:publish`, e.g. [`sdk:nodejs:publish`](https://github.com/dagger/dagger/pull/3809/files#diff-403f05c48ac8fbf799afe4dcf233f44d126ed2546723938ac76b1e30cbb1c75cR54-R65)

Follow-up:
- version current Homebrew bottled formula (`dagger.rb` -> `dagger@0.2.rb`)
- version current Homebrew tap formula (`dagger.rb` -> `dagger@0.2.rb`) - it's OK to inconvenience the few users that depend on this (talked to @aluzzardi about it)
- publish

**Comment by sipsma** [comment]
> related but from the distance. The CLI will eventually use the CLI helper, but for the moment, according to what we spoke with @sipsma and @aluzzardi we'll keep using the legacy buildkit provisioner since it makes more sense at this stage.

I think this is coming to a head already, e.g. [there's a lot of issues around building the shim on the fly](https://github.com/dagger/dagger/issues/3733) that might be most easily addressed by just including the shim in the engine image (or by making the shim the same thing as the session router binary). But if we make that change then the CLI can no longer depend on the legacy-buildkitd provisioner and has to use the new engine image.

---

The problem is the weird circular dependency we discussed where the CLI is a component of engine but will need to hardcode a dependency on an engine image (similar to Go SDK) in order to use the new provisioning.

Possibilities:
1. Move the CLI out to be its own submodule, either sibling to the Go SDK or possibly a submodule with a dependency on the Go SDK.
2. Just accept the weird circular dependency for now
   * This would mean that we hardcode the engine image ref in the CLI code the same as we do in our SDKs. Similarly, we update it when there is a new engine release.
   * To avoid too many circular deps, we just need to not put the CLI inside the engine image and then configure our workflows to not build a new engine image when only the CLI code was updated
3. Don't change anything, keep using legacy-buildkitd provisioner
   * We can't fix the shim build process (or make any other changes that would require anything in our custom engine image to be present, e.g. CNI support, wasm interpreters, etc.)

`2` is weird and kind of confusing, but afaict it would work for now? And also avoid the huge headache of managing the CLI as a go submodule the same way we do the Go SDK. Personally, if it is indeed workable, I think we should just do that now so that we aren't blocked on making changes like the one above to the shim.

However, if we are willing to accept the pain of `1`, then it feels like the most "logical" approach in the long run.

cc @aluzzardi @marcosnils let me know what you think, only change from our last discussion is that there are already compelling needs for more engine image customization.

**Comment by marcosnils** [comment]
> cc @aluzzardi @marcosnils let me know what you think, only change from our last discussion is that there are already compelling needs for more engine image customization.

If the image customization cases are becoming more important in a short time span then I'd say option `2` seems like the best one initially. Even though I don't like the circular spiral as well, option `3` involves some complexity that I'm not quite sure that it's worth putting the effort  ATM. @aluzzardi thoughts?

File: .github/workflows/publish-cli.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
+name: "publish-cli"
+
+# Only a single job with this concurrency can run at any given time
+concurrency: publish-cli
+
+on:
+  workflow_dispatch:
```
Comment (aluzzardi) [inline .github/workflows/publish-cli.yml:7]: @gerhard  shouldn't this be merged with `publish-engine` since they share the same tag?

File: cmd/dagger/exec.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+var execCmd = &cobra.Command{
+	Use:                   "exec [command]",
+	Aliases:               []string{"x"},
+	DisableFlagsInUseLine: true,
+	Long:                  "Executes the specified command in a Dagger session\n\nDAGGER_HOST will be injected automatically.",
+	Short:                 "Executes a command in a Dagger session",
+	Example: `
+dagger query <<EOF
```
Comment (aluzzardi) [inline cmd/dagger/exec.go:14]: wrong example :)

File: cmd/dagger/main.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+Global Flags:
+{{.InheritedFlags.FlagUsages | trimTrailingWhitespaces}}{{end}}{{if .HasHelpSubCommands}}
+
+Additional help topics:{{range .Commands}}{{if .IsAdditionalHelpTopicCommand}}
+  {{rpad .CommandPath .CommandPathPadding}} {{.Short}}{{end}}{{end}}{{end}}{{if .HasAvailableSubCommands}}
+
+Use "{{.CommandPath}} [command] --help" for more information about a command.{{end}}
+`)
+
+	queryCmd.Flags().StringVarP(&queryFile, "doc", "", "", "document query file")
```
Comment (aluzzardi) [inline cmd/dagger/main.go:65]: nit (for another time, I know this is from cloak): we should define flags where they are used rather than all in main

File: cmd/dagger/exec.go (hunk 4 above)
Comment (marcosnils) [reply cmd/dagger/exec.go:14]: Yep, copy/pasta. Still in progress

File: .github/workflows/publish-cli.yml (hunk 3 above)
Comment (marcosnils) [reply .github/workflows/publish-cli.yml:7]: Will be. Highlighted in TODO's here: https://github.com/dagger/dagger/pull/3748#issuecomment-1314074072

File: cmd/dagger/main.go (hunk 5 above)
Comment (marcosnils) [reply cmd/dagger/main.go:65]: Oh, I actually found it nicer to have them all in `main`. No strong opinions though, we can refactor this.

**Comment by shykes** [comment]
> The problem is the weird circular dependency we discussed where the CLI is a component of engine but will need to hardcode a dependency on an engine image (similar to Go SDK) in order to use the new provisioning.
> 
> Possibilities:
> 
> 1. Move the CLI out to be its own submodule, either sibling to the Go SDK or possibly a submodule with a dependency on the Go SDK.
> 2. Just accept the weird circular dependency for now
>    
>    * This would mean that we hardcode the engine image ref in the CLI code the same as we do in our SDKs. Similarly, we update it when there is a new engine release.
>    * To avoid too many circular deps, we just need to not put the CLI inside the engine image and then configure our workflows to not build a new engine image when only the CLI code was updated

This one I don't understand: since the CLI relies on the Go SDK, isn't all this logic already handled in the Go SDK? And if so, why would we need to hardcode the engine image ref in the CLI code?

> 3. Don't change anything, keep using legacy-buildkitd provisioner

This IMO is not an option. It's too confusing from a UX point of view.

> `2` is weird and kind of confusing, but afaict it would work for now? And also avoid the huge headache of managing the CLI as a go submodule the same way we do the Go SDK. Personally, if it is indeed workable, I think we should just do that now so that we aren't blocked on making changes like the one above to the shim.
> 
> However, if we are willing to accept the pain of `1`, then it feels like the most "logical" approach in the long run.

There is also an option 4: move the CLI under the Go SDK, since it is essentially a CLI version of that SDK.

It's not a perfect option, but it's very close to option 2 in its imperfections. So we should at least add it to the list and discuss it.

**Comment by sipsma** [comment]
> This one I don't understand: since the CLI relies on the Go SDK, isn't all this logic already handled in the Go SDK? And if so, why would we need to hardcode the engine image ref in the CLI code?

The current implementation and the one I believe we were going to go with (correct me if wrong @marcosnils ) is that the CLI would *not* rely on the Go SDK. It will just use `engine.Start` directly and thus bypass the SDK (but also by-passing the provisioning from the engine-session binary in the process).

> There is also an option 4: move the CLI under the Go SDK, since it is essentially a CLI version of that SDK.

Yeah this is a variation on option `1` in my mind. Basically, move the CLI out of the engine code directly and have it either be inside the Go SDK code base or another submodule that has a dependency on the Go SDK. Personally I'd be okay with either, it's more just which one is a lesser headache right now if we go this route (though both are probably pretty big headaches in terms of release and version management).

File: cmd/dagger/main.go (hunk 5 above)
Comment (gerhard) [reply cmd/dagger/main.go:66]: That's exactly what I mentioned yesterday.

I've seen the worst of what happens when this grows over the years, and while I am sure that this situation will be different, I would much rather see command flags living with the command source than everything in `main`. We don't need to do it in this PR - keep 'em small, keep 'em flowing - but we should address it within the next few weeks.

File: cmd/dagger/query.go (hunk 2 above)
Comment (gerhard) [reply cmd/dagger/query.go:119]: `+1`

**Comment by marcosnils** [comment]
> The current implementation and the one I believe we were going to go with (correct me if wrong @marcosnils ) is that the CLI would _not_ rely on the Go SDK. It will just use `engine.Start` directly and thus bypass the SDK (but also by-passing the provisioning from the engine-session binary in the process).

@sipsma since your last comment I've refactored the code to **only** use the SDK so we don't rely on the legacy mode anymore. Let's a have a quick chat today to hopefully come to a unified strategy here

File: cmd/dagger/main.go (hunk 5 above)
Comment (marcosnils) [reply cmd/dagger/main.go]: fixed @gerhard. It wasn't a big change at all, it's ok to scope it to this PR.

**Comment by marcosnils** [comment]
CLI  code and UX in this PR is ready for review @gerhard @aluzzardi @sipsma @shykes . 

After we decide where the CLI will live we need to finalize the publish specific tasks.

File: cmd/dagger/query.go (hunk 2 above)
Comment (marcosnils) [reply cmd/dagger/query.go:119]: working on this RN

File: cmd/dagger/query.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	result, err := json.MarshalIndent(res, "", "    ")
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// TODO use generics?
+func getKVInput(kvs map[string]string) map[string]interface{} {
```
Comment (sipsma) [inline cmd/dagger/query.go:126]: Personally I think I'd rather go with a fallback option until we can implement this with full graphql type introspection. IMO there's just too many gotchas remaining even though I totally agree this covers the majority of cases (e.g. "true" is always coerced to be a bool, any string containing `,` is a list, etc.).

One fallback option would be to just have users provide vars in the form of a json blob, e.g.

```
dagger query --vars '{"foo": "bar", "baz": ["1", "true"]}' <<EOF 
{
 query Blah($foo: String!, $baz: [String!]!) {
     ...
  }
}
EOF
```

I realize that is a much worse ux but it seems like something we could start with and then layer the nicer `--set foo=bar` stuff on top once we have the implementation that uses graphql type introspection ready.

Just my opinion, not obvious at all. If we do decide to go this route then I think we should at least tighten up a few parts, such as only interpreting a bar as a list if it's wrapped in `[` and `]`.

File: cmd/dagger/exec.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+			os.Exit(1)
+		}
+		listening <- fmt.Sprintf("%d", l.Addr().(*net.TCPAddr).Port)
+		http.Serve(l, nil)
+	}()
+
+	listenPort := <-listening
+	os.Setenv("DAGGER_SESSION_PORT", listenPort)
```
[the remark this answers is not in the document]
Comment (marcosnils) [reply cmd/dagger/exec.go:49]: This shouldn't affect SDK's since the idea of setting `DAGGER_SESSION_PORT` is so the wrapped process can use it to connect to `localhost:$PORT` to reach the API. For example: 
```
dagger exec -- sh -c 'curl -v \
-H "content-type:application/json" \
-d "{\"query\":\"{container{id}}\"}" \
http://localhost:$DAGGER_SESSION_PORT/query'
```

File: cmd/dagger/query.go (hunk 6 above)
Comment (marcosnils) [reply cmd/dagger/query.go:126]: > One fallback option would be to just have users provide vars in the form of a json blob, e.g.

Yep, I agree that JSON blobs would be a much better fallback mechanism. Let's ask @shykes what he thinks about.

File: cmd/dagger/exec.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sipsma) [reply cmd/dagger/exec.go:49]: Oh my understanding was that `dagger exec` was meant to wrap any command that uses a dagger sdk (so that e.g. everything wrapped by the command would be in the same session) as opposed to being meant to wrap literally anything, but I may be out of sync with the latest ideas around this.

**Comment by marcosnils** [comment]
Just adding some conclusions after a chat with @shykes and @aluzzardi 

1 - The CLI will live alongside the engine as drafted in this PR. We still need to polish some things around its deployment. More info about that here (https://github.com/dagger/dagger/issues/3830#issuecomment-1316288485).
2 - We've added `DAGGER_SESSION_URL` as a companion of `DAGGER_SESSION_PORT`  mostly to improve thx UX around DSI.
3 - `dagger query` will have two flags, `--var` and `--var-json` respectively to define arguments for the graphql query.

**Comment by marcosnils** [comment]
The CLI part of this PR is ready to :eye: 

I'll open a new PR for the distribution changes. @gerhard @shykes @sipsma mind giving one last :eye: please?

**Comment by shykes** [comment]
@marcosnils my only outstanding topic is the interface for `dagger exec` (what variables does it set, and how do we document it). Since we don't have a resolution on #3830 yet, we should decouple (ie. we should ship something for `dagger exec` that does not restrict our options for DSI, but doesn't break when DSI is defined either.

I suggest:

1. `dagger exec` sets `$DAGGER_SESSION_URL` as a convenience. It is guaranteed to always be an http URL. This is *not the same thing as DSI*, just a convenience
2. `dagger exec` *does not* set `$DAGGER_SESSION_PORT`, to avoid confusion and interference with DSI
3. When DSI is defined, we can make follow-up modifications to `dagger exec` to be compliant (all SDKs will need to be modified anyway).

WDYT @marcosnils @aluzzardi @sipsma ?

**Comment by marcosnils** [comment]
> WDYT @marcosnils @aluzzardi @sipsma ?

SGTM Solomon. just removed `DAGGER_SESSION_PORT` since I agree it shouldn't be there until DSI is initially defined.

**Review by sipsma — APPROVED** [review]
Looks good! The change with `--var` and `--var-json` SGTM, seems like a good compromise for now.
